lib/galaxy/jobs/__init__.py

```diff
@@ -1758,9 +1743,14 @@ def setup_external_metadata(self, exec_dir=None, tmp_dir=None,
         if datatypes_config is None:
             datatypes_config = os.path.join(self.working_directory, 'registry.xml')
             self.app.datatypes_registry.to_xml_file(path=datatypes_config)
-        command = self.external_output_metadata.setup_external_metadata([output_dataset_assoc.dataset for
-                                                                         output_dataset_assoc in
-                                                                         job.output_datasets + job.output_library_datasets],
+
+        output_datasets = {}
+        for output_dataset_assoc in job.output_datasets + job.output_library_datasets:
+            output_name = output_dataset_assoc.name
+            assert output_name not in output_datasets
+            output_datasets[output_dataset_assoc.name] = output_dataset_assoc.dataset
```

Comment:
Is `.name` guaranteed to be unique ? If not I think this would break things. My gut feeling is it's not unique. Would it hurt to `enumerate` here for the dictionary keys ?

Reply:
So this is the name of the tool output (like the name in the ``data`` tag - ``<data format="txt"  name="output">`` ), not the name of the dataset. I'm really hoping it is unique because I'm keying on it over and over downstream. Could a tool really have two outputs with the same output name? I've already started keying on it for like ToolProvidedMetadata (galaxy.json) in recent releases of Galaxy - because it is a lot more easy to read and write consistently than the dataset id - which requires transferring knowledge about the runtime of the output context to the tool.

Reply:
oh, missed that this isn't a HDA. Yes, that's a totally reasonable assumption.

Reply:
Extremely easy to miss given the misleading name.